### PR TITLE
ESI: Replace --disable-esi build option

### PR DESCRIPTION
**Pull request by yadij**
What this build option did can also be performed using
both options "--without-expat --without-xml2".

Check libexpat and libxml2 with SQUID_CHECK_LIB_WORKS.
Expand the macro to auto-define make variable
ENABLE_LIBFOO and compiler variable HAVE_LIBFOO
for working libraries.

Existing libraries checked with this macro do not need those
variables, but many others yet to be converted will.

Also, require pkg-check for library detection. Removing
the manual check for 'main()' function in the library ABI.

### CODE REVIEW

**Review by kinkie — APPROVED** [review]
Approving but having the extract of a test run attached to the PR would make me more confident

File: configure.ac [hunk 1]
```diff
@@ -839,27 +839,17 @@ AC_ARG_ENABLE(esi,
...
-      LIBEXPAT_LIBS="$LIBEXPAT_PATH $LIBEXPAT_LIBS"
-      AC_DEFINE(HAVE_LIBEXPAT,1,[Define to 1 if you have the expat library])
-    ],
-    [test "x$with_expat" = "xyes"],[AC_MSG_ERROR([Required library expat not found.])],
-    [AC_MSG_NOTICE([Library expat not found.])]
-  )
 ])
-AM_CONDITIONAL(ENABLE_LIBEXPAT,[test "x$LIBEXPAT_LIBS" != "x"])
-AC_SUBST(LIBEXPAT_LIBS)
+AS_IF([test "x$with_expat" = "xyes"],[squid_opt_use_esi=yes])
```
[the remark this answers is not in the document]
Comment (yadij) [reply configure.ac:852]: The new check macro explicitly sets `with_foo` to "no" unless the library is both found and working.

Since I have expanded scope to include libxml2 and simplified the final `AS_IF` now this `squid_opt_use_esi` has gone away.

File: acinclude/squid-util.m4 [hunk 2]
```diff
@@ -222,20 +222,23 @@ dnl 2) necessary library checks (to be executed by this macro unless the use of
...
 AC_DEFUN([SQUID_CHECK_LIB_WORKS],[
+AH_TEMPLATE(m4_toupper(m4_translit([HAVE_LIB$1], [-+.], [___])),[Define as 1 to enable '$1' library support.])
 AS_IF([m4_translit([test "x$with_$1" != "xno"], [-+.], [___])],[
   $2
   AS_IF([! test -z m4_toupper(m4_translit(["$LIB$1_LIBS"], [-+.], [___]))],[
     m4_toupper(m4_translit([CPPFLAGS="$LIB$1_CFLAGS $CPPFLAGS"], [-+.], [___]))
     m4_toupper(m4_translit([LIB$1_LIBS="$LIB$1_PATH $LIB$1_LIBS"], [-+.], [___]))
     AC_MSG_NOTICE([Library '$1' support: m4_translit([${with_$1:=yes (auto)} m4_toupper($LIB$1_LIBS)], [-+.], [___])])
     m4_translit([with_$1], [-+.], [___])=yes
+    AC_DEFINE(m4_toupper(m4_translit([HAVE_LIB$1], [-+.], [___])),1,[Define as 1 enable '$1' library support.])
```
Comment (rousskov) [inline acinclude/squid-util.m4:233]: Typo:

```suggestion
    AC_DEFINE(m4_toupper(m4_translit([HAVE_LIB$1], [-+.], [___])),1,[Define as 1 to enable '$1' library support.])
```

File: src/esi/ExpatParser.cc [hunk 3]
```diff
@@ -10,7 +10,7 @@
 
 #include "squid.h"
 
-#if USE_SQUID_ESI && HAVE_LIBEXPAT
+#if HAVE_LIBEXPAT
```
Comment (rousskov) [inline src/esi/ExpatParser.cc:13]: Here and elsewhere in this PR, I am against removing USE_SQUID_ESI precondition (even if HAVE_LIBEXPAT implies USE_SQUID_ESI today). If some ESI code is using USE_SQUID_ESI, then all ESI code should be using USE_SQUID_ESI (to the extent possible).

**Review by rousskov — COMMENTED** [review]
FWIW, I am unable to review the substance of the primary PR changes. I hope they have been tested well.

File: src/esi/ExpatParser.cc (hunk 3 above)
Comment (yadij) [reply src/esi/ExpatParser.cc:13]: Since the removal of the custom ESI parser `USE_SQUID_ESI` has become redundant with the two library `HAVE_FOO` macros. I have now gone the whole way to full removal of the `--disable-esi` build option.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: doc/release-notes/release-7.sgml.in [hunk 4]
```diff
@@ -210,6 +210,9 @@ This section gives an account of those changes in three categories:
 <sect1>Removed options<label id="removedoptions">
 <p>
 <descrip>
+	<tag>--disable-esi</tag>
+	<p>Removed. Use options <em>--without-expat --without-xml2</em> instead.
```
[the remark this answers is not in the document]
Comment (yadij) [reply doc/release-notes/release-7.sgml.in:214]: FTR; we have several groups of libraries that have this behaviour (openssl vs gnutls), (mit-krb5 vs heimdal vs gss), (OpenLDAP vs Sun LDAP vs Mozilla LDAP). None of these require both an `--enable-foo --with-foo` options to build the feature.
 Only ESI breaks the pattern of having both `--with-foo` and `--enable-foo` to link against a library, with all the extra complexity involved with validating the broken `--enable-foo --without-foo` builds.

File: acinclude/squid-util.m4 [hunk 5]
```diff
@@ -238,7 +238,7 @@ AS_IF([m4_translit([test "x$with_$1" != "xno"], [-+.], [___])],[
     AC_MSG_NOTICE([Library '$1' support: no (auto)])
   ])
 ])
-AM_CONDITIONAL(m4_toupper(m4_translit([ENABLE_LIB$1], [-+.], [___])),[test "x$with_$1" != "xno"])
+AM_CONDITIONAL(m4_toupper(m4_translit([ENABLE_LIB$1],[-+.],[___])),m4_translit([test "x$with_$1" != "xno"],[-+.],[___]))
```
Comment (kinkie) [inline acinclude/squid-util.m4:241]: This change seems out of scope

Comment (yadij) [reply acinclude/squid-util.m4:241]: This is already in master, just bringing the branch up to sync.

**Comment by yadij** [comment]
Superseded by later PRs
